sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -69,25 +76,37 @@ void start() throws TransientKinesisException {
       shardsMap.put(checkpoint.getShardId(), createShardIterator(kinesis, checkpoint));
     }
     shardIteratorsMap = shardsMap.build();
-    executorService = Executors.newFixedThreadPool(shardIteratorsMap.size());
-    recordsQueue = new LinkedBlockingQueue<>(queueCapacityPerShard * shardIteratorsMap.size());
-    for (final ShardRecordsIterator shardRecordsIterator : shardIteratorsMap.values()) {
-      executorService.submit(new Runnable() {
-
-        @Override
-        public void run() {
-          readLoop(shardRecordsIterator);
-        }
-      });
+    if (!shardIteratorsMap.isEmpty()) {
+      recordsQueue = new ArrayBlockingQueue<>(queueCapacityPerShard * shardIteratorsMap.size());
+      for (final ShardRecordsIterator shardRecordsIterator : shardIteratorsMap.values()) {
+        numberOfRecordsInAQueueByShard.put(shardRecordsIterator.getShardId(), new AtomicInteger());
+        executorService.submit(new Runnable() {
+
+          @Override
+          public void run() {
+            readLoop(shardRecordsIterator);
+          }
+        });
+      }
+    } else {
+      recordsQueue = new LinkedBlockingQueue<>();
     }
   }
 
   private void readLoop(ShardRecordsIterator shardRecordsIterator) {
     while (poolOpened.get()) {
       try {
-        List<KinesisRecord> kinesisRecords = shardRecordsIterator.readNextBatch();
-        for (KinesisRecord kinesisRecord : kinesisRecords) {
-          recordsQueue.put(kinesisRecord);
+        try {
+          List<KinesisRecord> kinesisRecords = shardRecordsIterator.readNextBatch();
+          for (KinesisRecord kinesisRecord : kinesisRecords) {
+            recordsQueue.put(kinesisRecord);
+            numberOfRecordsInAQueueByShard.get(kinesisRecord.getShardId()).incrementAndGet();
+          }
+        } catch (KinesisShardClosedException e) {
+          LOG.info("Shard iterator is closed, finishing the read loop", e);
+          waitUntilAllShardRecordsRead(shardRecordsIterator);
```

Comment:
Why is it important to wait until all records are read from it - can't we start reading from the new shards right away, and let the current thread complete via its natural course?

Reply:
It might cause other problems. If we process records from already closed shard and from its successor then we will have to handle checkpoint restoration differently. It seemed a more straightforward approach to just wait until records from a closed shard are processed and then start reading successors.